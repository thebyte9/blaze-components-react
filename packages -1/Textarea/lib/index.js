"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
exports.__esModule = true;
var utils_1 = __importDefault(require("@blaze-react/utils"));
var react_1 = __importStar(require("react"));
var Textarea = function (_a) {
    var value = _a.value, label = _a.label, limit = _a.limit, onChange = _a.onChange, error = _a.error, validationMessage = _a.validationMessage, required = _a.required, id = _a.id, _b = _a.utils, classNames = _b.classNames, ErrorMessage = _b.ErrorMessage, attrs = __rest(_a, ["value", "label", "limit", "onChange", "error", "validationMessage", "required", "id", "utils"]);
    var _c = react_1.useState(""), content = _c[0], setContent = _c[1];
    react_1.useEffect(function () {
        if (!content && value) {
            setContent(value);
        }
    }, []);
    var handleChange = function (event) {
        var newContent = event.target.value;
        if (limit && newContent.length > limit) {
            newContent = newContent.slice(0, limit);
        }
        setContent(newContent);
        onChange({ event: event, value: newContent });
    };
    var requiredClassName = classNames({ required: required });
    var total = !limit ? 0 : limit - content.length;
    return (react_1["default"].createElement("div", { className: "form-field form-field--textarea" },
        label && (react_1["default"].createElement("label", { htmlFor: id, className: requiredClassName }, label)),
        react_1["default"].createElement("textarea", __assign({ value: content, rows: attrs.rows, cols: attrs.cols, onChange: handleChange, required: required }, attrs)),
        !!limit && react_1["default"].createElement("span", null, total),
        error && react_1["default"].createElement(ErrorMessage, { message: validationMessage })));
};
Textarea.defaultProps = {
    error: false,
    label: "",
    limit: 0,
    placeholder: "",
    required: false,
    validationMessage: "This field is required",
    value: ""
};
exports["default"] = utils_1["default"](Textarea);
