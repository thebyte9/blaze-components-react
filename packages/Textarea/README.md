
## Description

The HTML textarea element represents a multi-line plain-text editing control, useful when you want to allow users to enter a sizeable amount of free-form text, for example a comment on a review or feedback form.

## Usage

* Simple

```js
<Textarea
    label="Simple textarea"
    placeholder="Content..."
    onChange={({event, value}) => {}}
    required />
```

* Limited

```js
<Textarea
  label="Textarea with limit"
  placeholder="Content..."
  onChange={({event, value}) => {}}
  value="lorem ipsum"
  limit={40} />
```

## API

##### Textarea can receive a number of `props` as follow:


| NAME   | TYPE | DEFAULT | 
| :---  | :---:  | :---: | 
| label | String | empty | 
| value | String | empty | 
| required | Boolean | false | 
| limit | Number | 0 | 
| onChange | Function | () => {} | 

