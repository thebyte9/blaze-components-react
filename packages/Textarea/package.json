{
  "name": "@blaze-react/text-area",
  "version": "0.6.0",
  "description": "",
  "main": "./dist/main.js",
  "scripts": {
    "build": "NODE_ENV=production webpack --config webpack.config.js",
    "build-storybook": "build-storybook -c .storybook -o docs-build",
    "lint": "eslint '*/**/*.{ts,tsx}' --ignore-path ../../.eslintignore --quiet --fix",
    "test": "jest",
    "test:coverage": "jest --coverage",
    "pre-publish": "yarn build && yarn lint && yarn test"
  },
  "keywords": [],
  "author": "byte9",
  "license": "MIT",
  "dependencies": {
    "@blaze-react/utils": "latest",
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  },
  "devDependencies": {
    "@blaze-react/blaze-components-theme": "^0.6.0",
    "webpack": "^5.29.0",
    "webpack-cli": "^4.6.0"
  }
}
