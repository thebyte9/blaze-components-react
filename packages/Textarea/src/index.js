import React, { Fragment, useState } from 'react';
import PropTypes from 'prop-types';

const Textarea = ({ value, label, limit, onChange, required, ...attrs }) => {
  const [content, setContent] = useState(value);

  const handleChange = event => {
    let _content = event.target.value;

    if (limit && _content.length > limit) _content = _content.slice(0, limit);

    setContent(_content);
    onChange({ event, value: _content });
  };

  const isRequired = required ? 'required' : '';
  const total = limit && limit - content.length;

  return (
    <Fragment>
      {label && (
        <label htmlFor={attrs.id} className={isRequired}>
          {label}
        </label>
      )}
      <textarea
        value={content}
        rows={attrs.rows}
        cols={attrs.cols}
        onChange={handleChange}
        required={required}
        {...attrs}
      />
      {!!limit && <span>{total}</span>}
    </Fragment>
  );
};

Textarea.propTypes = {
  label: PropTypes.string,
  value: PropTypes.string,
  required: PropTypes.bool,
  limit: PropTypes.number,
  onChange: PropTypes.func
};

Textarea.defaultProps = {
  label: '',
  value: '',
  required: false,
  limit: 0,
  onChange: () => {}
};

export default Textarea;
