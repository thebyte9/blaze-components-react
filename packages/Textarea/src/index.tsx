import React, {
  Fragment,
  FunctionComponent,
  TextareaHTMLAttributes,
  useState
} from "react";
interface ITextareaProps extends TextareaHTMLAttributes<HTMLTextAreaElement> {
  label: string;
  required?: boolean;
  limit?: number;
  onChange: (...args: any[]) => void;
  value?: string;
  placeholder?: string;
}
const Textarea: FunctionComponent<ITextareaProps> = ({
  value,
  label,
  limit,
  onChange,
  required,
  id,
  ...attrs
}) => {
  const [content, setContent] = useState<string>('');

  const handleChange = (event: React.ChangeEvent<HTMLTextAreaElement>): void => {
    let { target: { value: newContent } } = event;

    if (limit && newContent.length > limit) {
      newContent = newContent.slice(0, limit);
    }

    setContent(newContent);
    onChange({ event, value: newContent });
  };

  const requiredClassName = 'required';

  const isRequired: string = required ? requiredClassName : '';

  const getTotal = (): number => {
    if (!limit) {
      return 0;
    }
    return limit - content.length;
  }

  const total: number = getTotal();

  return (
    <Fragment>
      {label && (
        <label htmlFor={id} className={isRequired}>
          {label}
        </label>
      )}
      <textarea
        value={content}
        rows={attrs.rows}
        cols={attrs.cols}
        onChange={handleChange}
        required={required}
        {...attrs}
      />
      {!!limit && <span>{total}</span>}
    </Fragment>
  );
};
Textarea.defaultProps = {
  label: '',
  limit: 0,
  placeholder: '',
  required: false,
  value: ''
};
export default Textarea;
