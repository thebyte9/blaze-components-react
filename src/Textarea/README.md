
## Description

The HTML textarea element represents a multi-line plain-text editing control, useful when you want to allow users to enter a sizeable amount of free-form text, for example a comment on a review or feedback form.

## Usage

* Simple

```js
<Textarea
    label="Simple textarea"
    placeholder="Content..."
    onChange={({event, value}) => {}}
    required />
```

* Limited

```js
<Textarea
  label="Textarea with limit"
  placeholder="Content..."
  onChange={({event, value}) => {}}
  value="lorem ipsum"
  limit={40} />
```

## API

##### Textarea can receive a number of `props` as follow:


| NAME   | TYPE | DEFAULT | Description |
| :---  | :---:  | :---: | ------- |
| label | String | ''      | Does the textarea have a label? |
| value | String | ''  | Current value of the textarea |
| required | Boolean | false      | Is this a required field in the form? |
| limit | Number | 0      | Is there a maximum limit for the length of the value? |
| onChange | Function | () => {}     | returns callback with **event** and **value** |

