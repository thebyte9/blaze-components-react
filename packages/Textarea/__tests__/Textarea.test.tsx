import React from 'react';
import expect from 'expect';
import { shallow } from 'enzyme';
import Textarea from '../src';

describe('Textarea component', () => {
  test('should be defined and renders correctly (snapshot)', () => {
    const wrapper = shallow(<Textarea />);
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  test('should change textarea content', () => {
    const wrapper = shallow(<Textarea label="Simple extarea" required />);

    wrapper.find('textarea').simulate('change', { target: { value: 'byte9' } });
    expect(wrapper.find('textarea').prop('value')).toBe('byte9');
  });

  test('content length should be equal to limit', () => {
    const wrapper = shallow(<Textarea label="Simple extarea" limit={11} />);

    wrapper
      .find('textarea')
      .simulate('change', { target: { value: 'lorem ipsum dolor sit amet' } });
    expect(wrapper.find('textarea').prop('value')).toBe('lorem ipsum');

    const total = wrapper.find('span');
    expect(total.text()).toContain(0);
  });
});
