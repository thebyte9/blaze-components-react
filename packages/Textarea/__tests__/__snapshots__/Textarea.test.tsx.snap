// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea component should be defined and renders correctly (snapshot) 1`] = `
<Textarea
  error={false}
  label="Simple extarea"
  limit={0}
  onChange={[Function]}
  placeholder=""
  required={false}
  utils={
    Object {
      "ErrorMessage": [Function],
      "buildClassNames": [Function],
      "removeExtraSpaces": [Function],
      "uniqueId": [Function],
    }
  }
  validationMessage="This field is required"
  value=""
/>
`;
